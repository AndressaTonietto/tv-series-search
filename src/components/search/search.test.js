import "@testing-library/jest-dom";

import { fireEvent, render, screen } from "@testing-library/react";

import { BrowserRouter as Router } from "react-router-dom";
import Search from "./search";
import renderer from "react-test-renderer";
import useDebounce from "../../hooks/useDebounce";
import useFetchTvSeries from "../../hooks/useFetchSeries";

jest.mock("../../hooks/useDebounce");
jest.mock("../../hooks/useFetchSeries");

const testErrorMessage = "test error message";

describe("Search", () => {
  it("renders correctly", () => {
    useDebounce.mockReturnValue("");
    useFetchTvSeries.mockReturnValue({
      data: [],
      loading: false,
      error: null,
    });

    const tree = renderer
      .create(
        <Router>
          <Search />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays loading state", () => {
    useDebounce.mockReturnValue("");
    useFetchTvSeries.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    render(
      <Router>
        <Search />
      </Router>
    );
    expect(screen.getByTestId("loading-spinner")).toBeInTheDocument();
  });

  it("displays error state", () => {
    useDebounce.mockReturnValue("");
    useFetchTvSeries.mockReturnValue({
      data: null,
      loading: false,
      error: testErrorMessage,
    });

    render(
      <Router>
        <Search />
      </Router>
    );
    expect(
      screen.getByText((content, element) => content.includes(testErrorMessage))
    ).toBeInTheDocument();
  });

  it("displays no results state", () => {
    const noTvShowsMessage = "No tv shows matching your query were found";
    useDebounce.mockReturnValue("query");
    useFetchTvSeries.mockReturnValue({
      data: [],
      loading: false,
      error: null,
    });

    render(
      <Router>
        <Search />
      </Router>
    );
    const input = screen.getByPlaceholderText(/search/i);
    fireEvent.change(input, { target: { value: "query" } });
    expect(
      screen.getByText((content, element) => content.includes(noTvShowsMessage))
    ).toBeInTheDocument();
  });
});
