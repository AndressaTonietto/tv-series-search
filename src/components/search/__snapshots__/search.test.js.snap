// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  className="flex flex-col gap-4"
>
  <label
    className="input input-bordered flex items-center gap-2"
  >
    <input
      autoComplete="off"
      className="grow"
      id="search"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <svg
      className="w-4 h-4 opacity-70"
      fill="currentColor"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
        fillRule="evenodd"
      />
    </svg>
  </label>
  <div
    className="toast toast-center min-w-full"
  >
    <div
      className="alert alert-warning text-wrap"
    >
      No tv shows matching your query were found
    </div>
  </div>
  <div
    className="flex flex-col gap-4"
  />
</div>
`;
